---
title: "Content"
output:
  bookdown::html_document2:
    toc: true
    number_sections: true
---

<br>

```{r,child="assets/header-lab.Rmd"}
```

# Course objectives

In this course you will learn how to visualize your data in R.
In particular, you will learn:
 
- how to format the data necessary for ggplot 
- how to make bar-charts, box-plots and others using ggplot
- how to make PCA plots in ggplot
- how to make interactive plots in R using Rshiny
- how to host a Rshiny app in one of the available ser

# Page Info

This page contains links to different lectures (slides) and practical exercises (labs) that are part of this workshop. The links below are similar to that under [Schedule](home_schedule.html) but organised here by topic.

<div class="boxy boxy-lightbulb">
Input code blocks are displayed like shown below. The code language is displayed above the block. Shell scripts (**SH**) are to be executed in the linux terminal. **R** scripts are to be run in R either through the terminal, RGui or RStudio.

```{sh,eval=FALSE,block.title=TRUE}
command
```

<i class="fas fa-exclamation-circle"></i> Note &nbsp; <i class="fas fa-lightbulb"></i> Tip &nbsp; <i class="fas fa-comments"></i> Discuss &nbsp; <i class="fas fa-clipboard-list"></i> Task
</div>

<br>

This topic covers retrieving data needed for all the exercises in the entire course can be found in the following link. 

[<i class="fas fa-link"></i> <span class="large">Downloading data</span>](lab_download.html)

Some pre-course R warm-up below, if you fancy that:

[<i class="fas fa-link"></i> <span class="large">R warmup</span>](https://immuntech.github.io/Bioinformatics_Workshop_ImmTech/Data/2019-10-15/R_exercise_Lokesh.html)

# Introduction

Most of the analyses is carried out in R and it will be useful to learn some basics in R that are very much necessary for this course. Even if you have learnt these things already, it is good to freshen up your memory. 

[<i class="fas fa-link"></i> <span class="large">Introduction to R</span>](lab_r.html)  


# Main lab

## Data

In most of the exercises, we will use RNA-seq data (Illumina short reads) from the human A431 cell line. It is an epidermoid carcinoma cell line which is often used to study cancer and the cell cycle, and as a sort of positive control of epidermal growth factor receptor (EGFR) expression. A431 cells express very high levels of EGFR, in contrast to normal human fibroblasts.

The A431 cells were treated with gefinitib, which is an EGFR inhibitor and is used (under the trade name Iressa) as a drug to treat cancers with mutated and overactive EGFR. In the experiment, RNA was extracted at four time points: before the gefinitib treatment (t=0), and two, six and twenty-four hours after treatment (t=2, t=6, t=24, respectively), and sequenced using an Illumina HiSeq instrument in triplicates (thus there are 3x4=12 samples).

This data is part of the transcriptomics course that is also given by NBIS. We will use some of the `counts table` that was generated in the course after different transformations like manual filtering for low counts, VST and DESeq2. You don't have to know what these exactly mean to do these exercises :) These are basically the same data at different stages of the transcriptomics analysis.    

## ggplot part I

This section is for people who are not used to using `ggplot2` before! This section covers all the major basics to the grid graphics using `ggplot2`. These include some of the important functions in ggplot: geoms, colors and aesthetics, facets, barplots and errorbars.

[<i class="fas fa-link"></i> <span class="large">Basics in ggplot2</span>](lab_gg1.html)

## ggplot part II

This section is for the people who are comfortable in using in using `ggplot2` and would like to learn some advanced functions in `ggplot2`.

[<i class="fas fa-link"></i> <span class="large">Advanced ggplot2</span>](lab_gg_p2.html)

## Rshiny

This section is for the people here who are mainly interested in learning how to design the interactive `RShiny app`. This section covers all the basics of RShiny. The second section of this lab also includes some of the important functions which is important for mastering `Rshiny` 

[<i class="fas fa-link"></i> <span class="large">Rshiny</span>](lab_shiny.html)

## Rshiny Covid App

In this part, we learn how to write an app from scratch! So, we take the visualization of the covid data as an example from few countries that we would like to plot!

[<i class="fas fa-link"></i> <span class="large">Rshiny Covid App</span>](lab_shiny3.html)

# Solutions to exercises:

Here I have compiled all the solutions for the different exercises in each of the sections:

[<i class="fas fa-link"></i> <span class="large">Solutions</span>](solutions.html)

# Extra Tutorial for ggplot

In the following link, you can find several different ways to use `ggplot` and obtain incredible visualizations of the data. Depending on the kind of data you have and the kind of visualization you would like to see, you can follow it on the table contents in this following page:

[<i class="fas fa-link"></i> <span class="large">Top 50 ggplot2 Visualizations</span>](http://r-statistics.co/Top50-Ggplot2-Visualizations-MasterList-R-Code.html)

***
