<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Lokesh Mano • 02-Feb-2024" />


<title>Phylogenetic Trees</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="site_libs/font-awesome-6.4.2/fonts/fontawesome-webfont.ttf"/>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
#TOC {
  display: none !important;
}
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img src="assets/logo.svg" id="logo" style="height:22px;margin:0;"/></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="home_schedule.html">Schedule</a>
</li>
<li>
  <a href="home_content.html">Content</a>
</li>
<li>
  <a href="home_precourse.html">Precourse</a>
</li>
<li>
  <a href="home_info.html">Info</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Phylogenetic Trees</h1>
<h3 class="subtitle">Workshop on Data Visualization in R</h3>
<h4 class="author"><b>Lokesh Mano</b> • 02-Feb-2024</h4>

</div>


<!-- rmd lab header -->
<p><br></p>
<p>For this course, I have used one of my very simple tree from a few different organisms within the Archaea superkingdom. More specifically, it is the concatenated ribosomal protein tree within archaea, with representatives from the different Phyla: Asgard archaea (Thorarchaeota, Odinarchaeota, Lokiarchaeota and Heimdallarchaeota), TACK, Euryarchaeota and DPANN. The maximum likelihood phylogeny was reconstructed in IQ-TREE with the LG+F+I+G4 model. The alignment was trimmed to 6,732 positions with the BMGE (Block Mapping and Gathering with Entropy) tool.</p>
<div id="ggtree" class="section level1" number="1">
<h1><span class="header-section-number">1</span> ggtree</h1>
<p><code>ggtree</code> is an R package that extends ggplot2 for visualizating and annotating phylogenetic trees with their covariates and other associated data. <code>ggtree</code> in combination with <code>treeio</code> supports several file formats, including:</p>
<ul>
<li><code>read.tree</code> for reading Newick files.</li>
<li><code>read.phylip</code> for reading Phylip files.</li>
<li><code>read.jplace</code> for reading Jplace files.</li>
<li><code>read.nhx</code> for reading NHX files.</li>
<li><code>read.beast</code> for parsing output of BEAST</li>
<li><code>read.codeml</code> for parsing output of CODEML (rst and mlc files)</li>
<li><code>read.codeml_mlc</code> for parsing mlc file (output of CODEML)</li>
<li><code>read.hyphy</code> for parsing output of HYPHY</li>
<li><code>read.jplace</code> for parsing jplace file including output from EPA and pplacer</li>
<li><code>read.nhx</code> for parsing NHX file including output from PHYLODOG and RevBayes</li>
<li><code>read.paml_rst</code> for parsing rst file (output of BASEML and CODEML)</li>
<li><code>read.r8s</code> for parsing output of r8s</li>
<li><code>read.raxml</code> for parsing output of RAxML</li>
</ul>
</div>
<div id="reading-treefile" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Reading treefile</h1>
<pre class="r"><code>library(ggtree)
library(treeio)

arch_tree &lt;- read.newick(&quot;data/arch_newick.txt&quot;)
arch_tree</code></pre>
<details>
<summary>
output
</summary>
<pre><code>## 
## Phylogenetic tree with 25 tips and 24 internal nodes.
## 
## Tip labels:
##   DPANN_Pace_RBG_13_36_9, DPANN_Woese_UBA94, EURY_Thermoplasma_volcanium_GSS1, Eury_Halobaculum_gomorrense, Eury_Methanothermococcus_okinawensis, TACK_Bathy_SMTZ80, ...
## Node labels:
##   100, 10.9, 100, 100, 72, 35.9, ...
## 
## Rooted; includes branch lengths.</code></pre>
</details>
<p>Just like with <code>ggplot2</code> we created a basic canvas with <code>ggplot(...)</code> and added layers with <code>+geom_???()</code>, we can do the same here. The <code>ggtree</code> package gives us a <code>geom_tree()</code> function. Because <code>ggtree</code> is built on top of <code>ggplot2</code>, you get <code>ggplot2</code>’s default gray theme with white lines. You can override this with a theme from the <code>ggtree</code> package.</p>
<p>Because you’ll almost always want to add a tree geom and remove the default background and axes, the <code>ggtree()</code> function is essentially a shortcut for <code>ggplot(...) + geom_tree() + theme_tree()</code>.</p>
<pre class="r"><code>ggtree(arch_tree)
#Same as:
#ggplot(arch_tree) + geom_tree() + theme_tree()</code></pre>
<details>
<summary>
plot
</summary>
<p><img src="lab_phylo_files/figure-html/unnamed-chunk-5-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
<p>One can also customize how the tree looks like, just as in <code>ggplot</code></p>
<pre class="r"><code>ggtree(arch_tree, branch.length=&quot;none&quot;, color=&quot;blue&quot;, size=2, linetype=3)</code></pre>
<details>
<summary>
plot
</summary>
<p><img src="lab_phylo_files/figure-html/unnamed-chunk-6-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
<p>You can also use different orientations to show the tree.</p>
<pre class="r"><code>ggtree(arch_tree, color=&quot;blue&quot;, layout = &quot;circular&quot;)</code></pre>
<details>
<summary>
plot
</summary>
<p><img src="lab_phylo_files/figure-html/unnamed-chunk-7-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
</div>
<div id="geoms-and-annotation" class="section level1" number="3">
<h1><span class="header-section-number">3</span> Geoms and annotation</h1>
<p>Just like in <code>ggplot()</code> you can add layers to the tree as <code>geoms</code>.</p>
<pre class="r"><code>p &lt;- ggtree(arch_tree)
p + geom_tiplab(color = &quot;blue&quot;) + theme_tree2()</code></pre>
<details>
<summary>
plot
</summary>
<p><img src="lab_phylo_files/figure-html/unnamed-chunk-8-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
<p>Before we can go further into annotating the tree, we need to understand how ggtree is handling the tree structure internally. Some of the functions in ggtree for annotating clades need a parameter specifying the internal node number. To get the internal node number, user can use <code>geom_text</code> to display it, where the label is an aesthetic mapping to the “node variable” stored inside the tree object (think of this like the continent variable inside the gapminder object).</p>
<pre class="r"><code>p + geom_text(aes(label=node), hjust=-.3)</code></pre>
<details>
<summary>
plot
</summary>
<p><img src="lab_phylo_files/figure-html/unnamed-chunk-9-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
<p>Another way to get the internal node number is using the function <code>MRCA()</code> which stands for Most Recent Common Ancestor. From the tree above, if you would like to get the number of the particular node that is common to all the “Lokiarchaea”. You can do so like below:</p>
<pre class="r"><code>MRCA(arch_tree, .node1 = &quot;Baja_Loki3&quot;, .node2 =  &quot;Loki_CR4&quot;)</code></pre>
<details>
<summary>
output
</summary>
<pre><code>## [1] 46</code></pre>
</details>
<div id="clade-labeling" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Clade labeling</h2>
<p>Now, let us see how we can use this information to make it scientifically intuitive.</p>
<pre class="r"><code>p + 
  geom_cladelabel(node=46, label=&quot;Lokiarchaeota&quot;, color=&quot;green&quot;)</code></pre>
<details>
<summary>
plot
</summary>
<p><img src="lab_phylo_files/figure-html/unnamed-chunk-11-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
<p>ALternatively, we can also highlight an entire clade:</p>
<pre class="r"><code>p + 
  geom_hilight(node=39, fill = &quot;purple&quot;) +
  geom_cladelabel(node=39, label=&quot;Thorarchaeota&quot;, color=&quot;purple&quot;) +
  geom_hilight(node=46, fill = &quot;green&quot;) +
  geom_cladelabel(node=46, label=&quot;Lokiarchaeota&quot;, color=&quot;green&quot;)</code></pre>
<details>
<summary>
plot
</summary>
<p><img src="lab_phylo_files/figure-html/unnamed-chunk-12-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
</div>
<div id="taxa-connections" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> Taxa connections</h2>
<p>You can make connections to represent processes like gene transfer events on the tree.</p>
<pre class="r"><code>p + 
  geom_hilight(node=30, fill = &quot;orange&quot;) +
  geom_cladelabel(node=30, label=&quot;Euryarchaeota&quot;, color=&quot;orange&quot;) +
  geom_hilight(node=46, fill = &quot;green&quot;) +
  geom_cladelabel(node=46, label=&quot;Lokiarchaeota&quot;, color=&quot;green&quot;) +
  geom_taxalink(&quot;Eury_Methanothermococcus_okinawensis&quot;, &quot;Loki_GC14&quot;, color = &quot;red&quot;)</code></pre>
<details>
<summary>
plot
</summary>
<p><img src="lab_phylo_files/figure-html/unnamed-chunk-13-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
</div>
<div id="group-clades-and-otus" class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> Group clades and OTUs</h2>
<p>You can group the different tips using the <code>groupOTU</code> function or you can group the tips based on a node using <code>groupClade</code> function. These groups you can then further use it for other geoms and so on.</p>
<p>Let us say, we want to show the Asgrad archaea separately from the other archaea in this tree. One way to this would be:</p>
<div id="groupotu" class="section level3" number="3.3.1">
<h3><span class="header-section-number">3.3.1</span> groupOTU</h3>
<pre class="r"><code>arch_tree &lt;- groupOTU(arch_tree, c(&quot;DPANN_Pace_RBG_13_36_9&quot;, &quot;DPANN_Woese_UBA94&quot;,
                                   &quot;EURY_Thermoplasma_volcanium_GSS1&quot;, &quot;Eury_Halobaculum_gomorrense&quot;, &quot;Eury_Methanothermococcus_okinawensis&quot;, 
                                   &quot;TACK_Bathy_SMTZ80&quot;, &quot;TACK_Nitrososphaera_viennensis&quot;, &quot;TACK_Sulfolobus_islandicus_LS215&quot;))

ggtree(arch_tree, aes(color=group)) + geom_tiplab()</code></pre>
<details>
<summary>
plot
</summary>
<p><img src="lab_phylo_files/figure-html/unnamed-chunk-14-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
<p>We can also get more detailed:</p>
<pre class="r"><code> cls &lt;- list(Heimdallarchaeota = c(&quot;Heim_AB125&quot;, &quot;Heim_LC2&quot;, &quot;Heim_LC3&quot;), 
             Thorarchaeota = c(&quot;Baja_Thor&quot;, &quot;Thor_AB25&quot;, &quot;Thor_MP11T&quot;, &quot;Thor_SMTZ45&quot;, &quot;Thor_SMTZ_145&quot;, &quot;Thor_MP8T&quot;, &quot;Thor_MP9T&quot;, &quot;Thor_SMTZ_183&quot;), 
             Lokiarchaeota = c(&quot;Loki_CR4&quot;, &quot;Loki_GC14&quot;, &quot;Baja_Loki2&quot;, &quot;Baja_Loki1&quot;, &quot;Baja_Loki3&quot;),
             Odinarchaeota = &quot;Odin_LCB4&quot;)

arch_tree &lt;- groupOTU(arch_tree, cls)
ggtree(arch_tree) + geom_tiplab(aes(color=group)) + 
  scale_color_manual(values=c(&quot;black&quot;, rainbow(4)), labels = c(&quot;OutGroup&quot;, names(cls)))</code></pre>
<details>
<summary>
plot
</summary>
<p><img src="lab_phylo_files/figure-html/unnamed-chunk-15-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
<p>One thing to notice hear that you can group tips from different clades together as well.</p>
<pre class="r"><code>baja &lt;- list(Baja = c(&quot;Baja_Thor&quot;, &quot;Baja_Loki2&quot;, 
                      &quot;Baja_Loki1&quot;, &quot;Baja_Loki3&quot;))

arch_tree &lt;- groupOTU(arch_tree, baja)
ggtree(arch_tree) + geom_tiplab(aes(color=group)) + 
  scale_color_manual(values=c(&quot;black&quot;, &quot;blue&quot;), 
                     labels = c(&quot;Public genomes&quot;, &quot;Baja genomes&quot;))</code></pre>
<details>
<summary>
plot
</summary>
<p><img src="lab_phylo_files/figure-html/unnamed-chunk-16-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
</div>
<div id="groupclade" class="section level3" number="3.3.2">
<h3><span class="header-section-number">3.3.2</span> groupClade</h3>
<p>Now, if we are only interested in showing the differences based on the clades, we can use <code>groupClade</code>.</p>
<pre class="r"><code>arch_tree &lt;- groupClade(arch_tree, c(39, 46, 35, 27))

ggtree(arch_tree) + geom_tiplab(aes(color=group)) + 
  scale_color_manual(values=c(rainbow(4), &quot;black&quot;), 
                     labels = c(&quot;Odin&quot;, &quot;Thor&quot;, &quot;Loki&quot;, &quot;Heimdall&quot;, &quot;OutGroup&quot;))</code></pre>
<details>
<summary>
plot
</summary>
<p><img src="lab_phylo_files/figure-html/unnamed-chunk-17-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
</div>
</div>
<div id="showing-bootstrap-values" class="section level2" number="3.4">
<h2><span class="header-section-number">3.4</span> Showing bootstrap values</h2>
<p>If you notice the tree object, it actually contains <code>node.label</code> that are bootstrap values.</p>
<pre class="r"><code>arch_tree</code></pre>
<details>
<summary>
output
</summary>
<pre><code>## 
## Phylogenetic tree with 25 tips and 24 internal nodes.
## 
## Tip labels:
##   DPANN_Pace_RBG_13_36_9, DPANN_Woese_UBA94, EURY_Thermoplasma_volcanium_GSS1, Eury_Halobaculum_gomorrense, Eury_Methanothermococcus_okinawensis, TACK_Bathy_SMTZ80, ...
## Node labels:
##   100, 10.9, 100, 100, 72, 35.9, ...
## 
## Rooted; includes branch lengths.</code></pre>
</details>
<p>There are different ways we could show this values on the tree. For example,</p>
<pre class="r"><code>ggtree(arch_tree) + geom_nodelab() + geom_tiplab() + theme_tree()</code></pre>
<details>
<summary>
plot
</summary>
<p><img src="lab_phylo_files/figure-html/unnamed-chunk-19-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
<p>Now, let us say that you only want to show the values that are above 80, then you would have to do some trick here, like:</p>
<pre class="r"><code>ggtree(arch_tree) + 
  geom_nodepoint(aes(subset = as.numeric(label) &gt; 80), size = 3) + 
  geom_tiplab() + 
  theme_tree()</code></pre>
<details>
<summary>
plot
</summary>
<p><img src="lab_phylo_files/figure-html/unnamed-chunk-20-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
<p><i class="fas fa-exclamation-circle"></i> Note   Notice that the <code>node.label</code> are in string, so must remember to use <code>as.numeric()</code> option be able to filter them to show them.</p>
<p><i class="fas fa-exclamation-circle"></i> Note   If the tree does not have bootstrap values with in the treefile, one can add them separately as a <code>data.frame</code> using <code>%&lt;+%</code> function, as shown in the <strong>Adding data to tree</strong> section below.</p>
</div>
<div id="subsetting-trees" class="section level2" number="3.5">
<h2><span class="header-section-number">3.5</span> Subsetting trees</h2>
<p>It is possible to specific <code>tips</code> from the tree using <code>drop.tip()</code> fuction. Let us say, we want to remove <code>Thor_SMTZ45</code> and <code>Baja_Loki1</code>, becuase of the reason that their genomes are not close to complete. We can do that by:</p>
<pre class="r"><code>to_drop &lt;- c(&quot;Thor_SMTZ45&quot;, &quot;Baja_Loki1&quot;)
arch_tree_reduced &lt;- drop.tip(arch_tree, to_drop)
arch_tree_reduced &lt;- groupClade(arch_tree_reduced, c(37, 43, 33, 25))

ggtree(arch_tree_reduced) + geom_tiplab(aes(color=group)) + 
  scale_color_manual(values=c(rainbow(4), &quot;black&quot;), 
                     labels = c(&quot;Odin&quot;, &quot;Thor&quot;, &quot;Loki&quot;, &quot;Heimdall&quot;, &quot;OutGroup&quot;))</code></pre>
<details>
<summary>
plot
</summary>
<p><img src="lab_phylo_files/figure-html/unnamed-chunk-21-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
<p>One can also subset a tree, by using the internal node number as shown in the exercises earlier. We can do this by using the <code>tree_subset()</code> function:</p>
<pre class="r"><code>asg_tree &lt;- treeio::tree_subset(arch_tree, node=34, levels_back=0)
ggtree(asg_tree) + geom_tiplab() </code></pre>
<p><i class="fas fa-exclamation-circle"></i> Note   The branch lengths and all the other information from the tree are maintained in the new object, when remove tips or subset a tree.</p>
</div>
</div>
<div id="adding-data-to-tree" class="section level1" number="4">
<h1><span class="header-section-number">4</span> Adding data to tree</h1>
<p>In this part, let us try to see what kind of ways one could add metadata to a tree.</p>
<div id="tables" class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> Tables</h2>
<pre class="r"><code>env_info &lt;- read.table(&quot;data/tree_env.tsv&quot;, sep=&quot;\t&quot;, 
                stringsAsFactor=F, row.names = 1, header = T)

p &lt;- ggtree(arch_tree) + 
  geom_tiplab(aes(color=group)) +
  scale_color_manual(values=c(rainbow(4), &quot;black&quot;), 
    labels = c(&quot;Odin&quot;, &quot;Thor&quot;, &quot;Loki&quot;, &quot;Heimdall&quot;, &quot;OutGroup&quot;)) +
  theme(legend.position = &quot;none&quot;)

gheatmap(p, env_info, offset = 4, width=1.5, 
    colnames_angle = 90, colnames_offset_y = 5,
    legend_title = &quot;Environment sampled&quot;) </code></pre>
<details>
<summary>
plot
</summary>
<p><img src="lab_phylo_files/figure-html/unnamed-chunk-23-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
<p>Similarly we can plot a heatmap:</p>
<pre class="r"><code>env_hmap &lt;- read.table(&quot;data/tree_hmap.tsv&quot;, sep=&quot;\t&quot;, 
                stringsAsFactor=F, row.names = 1, header = T)

p &lt;- ggtree(arch_tree) + 
  geom_tiplab(aes(color=group)) +
  scale_color_manual(values=c(rainbow(4), &quot;black&quot;), 
    labels = c(&quot;Odin&quot;, &quot;Thor&quot;, &quot;Loki&quot;, &quot;Heimdall&quot;, &quot;OutGroup&quot;)) +
  theme(legend.position = &quot;none&quot;)

gheatmap(p, env_hmap, offset = 4, width=1.5, 
    colnames_angle = 90, colnames_offset_y = 5,
    legend_title = &quot;Environment sampled&quot;) </code></pre>
<details>
<summary>
plot
</summary>
<p><img src="lab_phylo_files/figure-html/unnamed-chunk-24-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
</div>
</div>
<div id="session-info" class="section level1" number="5">
<h1><span class="header-section-number">5</span> Session info</h1>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 4.1.3 (2022-03-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 22.04.3 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so
## 
## locale:
##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   
## 
## attached base packages:
## [1] grid      stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] treeio_1.27.0.002      ggtree_3.11.0          pheatmap_1.0.12       
##  [4] swemaps_1.0            mapdata_2.3.1          maps_3.4.2            
##  [7] gridExtra_2.3          jpeg_0.1-10            ggpubr_0.6.0          
## [10] cowplot_1.1.3          ggthemes_5.0.0         scales_1.3.0          
## [13] ggrepel_0.9.5          wesanderson_0.3.7      forcats_1.0.0         
## [16] stringr_1.5.1          purrr_1.0.2            readr_2.1.5           
## [19] tidyr_1.3.1            tibble_3.2.1           tidyverse_2.0.0       
## [22] reshape2_1.4.4         ggplot2_3.4.4          formattable_0.2.1     
## [25] kableExtra_1.4.0       dplyr_1.1.4            lubridate_1.9.3       
## [28] leaflet_2.2.1          yaml_2.3.8             fontawesome_0.5.2.9000
## [31] captioner_2.2.3        bookdown_0.37          knitr_1.45            
## 
## loaded via a namespace (and not attached):
##  [1] nlme_3.1-155            fs_1.6.3                RColorBrewer_1.1-3     
##  [4] tools_4.1.3             backports_1.4.1         bslib_0.6.1            
##  [7] utf8_1.2.4              R6_2.5.1                lazyeval_0.2.2         
## [10] mgcv_1.8-39             colorspace_2.1-0        withr_3.0.0            
## [13] tidyselect_1.2.0        compiler_4.1.3          cli_3.6.2              
## [16] xml2_1.3.6              labeling_0.4.3          sass_0.4.8             
## [19] yulab.utils_0.1.4       systemfonts_1.0.5       digest_0.6.34          
## [22] rmarkdown_2.25          svglite_2.1.3           pkgconfig_2.0.3        
## [25] htmltools_0.5.7         fastmap_1.1.1           highr_0.10             
## [28] htmlwidgets_1.6.4       rlang_1.1.3             rstudioapi_0.15.0      
## [31] gridGraphics_0.5-1      jquerylib_0.1.4         farver_2.1.1           
## [34] generics_0.1.3          jsonlite_1.8.8          crosstalk_1.2.1        
## [37] car_3.1-2               magrittr_2.0.3          ggplotify_0.1.2        
## [40] patchwork_1.2.0         Matrix_1.6-5            Rcpp_1.0.12            
## [43] munsell_0.5.0           fansi_1.0.6             ape_5.7-1              
## [46] abind_1.4-5             lifecycle_1.0.4         stringi_1.8.3          
## [49] carData_3.0-5           plyr_1.8.9              parallel_4.1.3         
## [52] lattice_0.20-45         splines_4.1.3           hms_1.1.3              
## [55] pillar_1.9.0            ggsignif_0.6.4          glue_1.7.0             
## [58] evaluate_0.23           ggfun_0.1.4             leaflet.providers_2.0.0
## [61] vctrs_0.6.5             tzdb_0.4.0              gtable_0.3.4           
## [64] cachem_1.0.8            xfun_0.41               broom_1.0.5            
## [67] tidytree_0.4.6          rstatix_0.7.2           viridisLite_0.4.2      
## [70] aplot_0.2.2             memoise_2.0.1           timechange_0.3.0       
## [73] ellipsis_0.3.2</code></pre>
<hr />
</div>

<footer class="footer">
<div class="container container-footer">
<div class="row">
<div class="col-sm-8" style="text-align:left;vertical-align:middle;">
<p style="color:#bdbdbd;">
<b>2023</b> • NBIS • SciLifeLab
</p>
</div>
<div class="col-sm-4" style="text-align:right;vertical-align:middle;">
<p style="color:#bdbdbd;">
<span> <span class="footericon" style="padding-right:4px; padding-left:4px"> <a href="https://nbis.se/"><svg aria-hidden="true" role="img" viewBox="0 0 496 512" style="height:1em;width:0.97em;vertical-align:-0.125em;margin-left:auto;margin-right:auto;font-size:inherit;fill:#bdbdbd;overflow:visible;position:relative;"><path d="M248 8C111.03 8 0 119.03 0 256s111.03 248 248 248 248-111.03 248-248S384.97 8 248 8zm82.29 357.6c-3.9 3.88-7.99 7.95-11.31 11.28-2.99 3-5.1 6.7-6.17 10.71-1.51 5.66-2.73 11.38-4.77 16.87l-17.39 46.85c-13.76 3-28 4.69-42.65 4.69v-27.38c1.69-12.62-7.64-36.26-22.63-51.25-6-6-9.37-14.14-9.37-22.63v-32.01c0-11.64-6.27-22.34-16.46-27.97-14.37-7.95-34.81-19.06-48.81-26.11-11.48-5.78-22.1-13.14-31.65-21.75l-.8-.72a114.792 114.792 0 0 1-18.06-20.74c-9.38-13.77-24.66-36.42-34.59-51.14 20.47-45.5 57.36-82.04 103.2-101.89l24.01 12.01C203.48 89.74 216 82.01 216 70.11v-11.3c7.99-1.29 16.12-2.11 24.39-2.42l28.3 28.3c6.25 6.25 6.25 16.38 0 22.63L264 112l-10.34 10.34c-3.12 3.12-3.12 8.19 0 11.31l4.69 4.69c3.12 3.12 3.12 8.19 0 11.31l-8 8a8.008 8.008 0 0 1-5.66 2.34h-8.99c-2.08 0-4.08.81-5.58 2.27l-9.92 9.65a8.008 8.008 0 0 0-1.58 9.31l15.59 31.19c2.66 5.32-1.21 11.58-7.15 11.58h-5.64c-1.93 0-3.79-.7-5.24-1.96l-9.28-8.06a16.017 16.017 0 0 0-15.55-3.1l-31.17 10.39a11.95 11.95 0 0 0-8.17 11.34c0 4.53 2.56 8.66 6.61 10.69l11.08 5.54c9.41 4.71 19.79 7.16 30.31 7.16s22.59 27.29 32 32h66.75c8.49 0 16.62 3.37 22.63 9.37l13.69 13.69a30.503 30.503 0 0 1 8.93 21.57 46.536 46.536 0 0 1-13.72 32.98zM417 274.25c-5.79-1.45-10.84-5-14.15-9.97l-17.98-26.97a23.97 23.97 0 0 1 0-26.62l19.59-29.38c2.32-3.47 5.5-6.29 9.24-8.15l12.98-6.49C440.2 193.59 448 223.87 448 256c0 8.67-.74 17.16-1.82 25.54L417 274.25z"></path></svg></a> </span> <span class="footericon" style="padding-right:4px; padding-left:4px"> <a href="https://twitter.com/NBISwe"><svg aria-hidden="true" role="img" viewBox="0 0 512 512" style="height:1em;width:1em;vertical-align:-0.125em;margin-left:auto;margin-right:auto;font-size:inherit;fill:#bdbdbd;overflow:visible;position:relative;"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"></path></svg></a> </span> <span class="footericon" style="padding-left:4px"> <a href="https://www.linkedin.com/company/nbisweden/"><svg aria-hidden="true" role="img" viewBox="0 0 448 512" style="height:1em;width:0.88em;vertical-align:-0.125em;margin-left:auto;margin-right:auto;font-size:inherit;fill:#bdbdbd;overflow:visible;position:relative;"><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"></path></svg></a> </span> </span>
</p>
</div>
</div>
</div>
</footer>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
