<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Lokesh Mano • 27-Apr-2022" />


<title>PCA and Heatmap</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="site_libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
#TOC {
  display: none !important;
}
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img src="assets/logo.svg" id="logo" style="height:22px;margin:0;"/></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="home_schedule.html">Schedule</a>
</li>
<li>
  <a href="home_content.html">Content</a>
</li>
<li>
  <a href="home_precourse.html">Precourse</a>
</li>
<li>
  <a href="home_info.html">Info</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">PCA and Heatmap</h1>
<h3 class="subtitle">Workshop on Data Visualization in R</h3>
<h4 class="author"><b>Lokesh Mano</b> • 27-Apr-2022</h4>

</div>


<!-- rmd lab header -->
<p><br></p>
<div id="pca" class="section level1">
<h1><span class="header-section-number">1</span> PCA</h1>
<p>Let us first make a PCA object. For this, we will use the <code>VST</code> data, because it makes sense to use the normalized data for building the PCA. To run PCA, we use the R function <code>prcomp()</code>. It takes in a matrix where samples are rows and variables are columns. Therefore, we transpose our count matrix using the function <code>t()</code>. If we do not transpose, then PCA is run on the genes rather than the samples.</p>
<pre class="r"><code>gc_vst &lt;- read.table(&quot;data/counts_vst.txt&quot;, header = T, row.names = 1, sep = &quot;\t&quot;)
vst_pca &lt;- prcomp(t(gc_vst))</code></pre>
<p>After you computer the PCA, if you type the object <code>vst_pca$</code> and press <code>TAB</code>, you will notice that this R object has multiple vecors and data.frames within it. Some of the important ones are</p>
<ul>
<li><code>sdev:</code> the standard deviations of the principal components</li>
<li><code>x:</code> the coordinates of the samples (observations) on the principal components.</li>
<li><code>rotation:</code> the matrix of variable loadings (columns are eigenvectors).</li>
</ul>
<p><i class="fas fa-exclamation-circle"></i> Note   There are quite a few functions in R from different packages that can run PCA. So, one should look into the structure of the PCA object and import it into <code>ggplot</code> accordingly!</p>
<div id="variance-of-components-scree-plot" class="section level2">
<h2><span class="header-section-number">1.1</span> Variance of components (Scree plot)</h2>
<p>First, let us look into plotting the variance explained by the top PCs.</p>
<pre class="r"><code>frac_var &lt;- function(x) x^2/sum(x^2)
library(scales)

vst_pca$sdev %&gt;% 
  as_tibble() %&gt;% 
  frac_var() %&gt;% 
  mutate(Comp = colnames(vst_pca$x)) %&gt;% 
  slice(1:9) %&gt;% 
  ggplot(aes(x=Comp, y = value)) + 
  geom_bar(stat = &quot;identity&quot;, fill = &quot;#4DC5F9&quot;) +
  geom_hline(yintercept = 0.03, linetype=2) +
  xlab(&quot;Principal Components&quot;) +
  scale_y_continuous(name = &quot;Variance Explained&quot;, breaks = seq(0,0.8,0.1), labels = percent_format(accuracy = 5L)) +
  theme_classic(base_size = 14)</code></pre>
<details>
<p><summary>plot</summary></p>
<p><img src="lab_pca_hmap_files/figure-html/unnamed-chunk-5-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
</div>
<div id="pca-plot" class="section level2">
<h2><span class="header-section-number">1.2</span> PCA plot</h2>
<p>So, looks like the first two components explain almost 85% of the data. Now, let us look into building a plot out of these components. From the above object, to get the scatter plot for the samples, we need to look into <code>vst_pca$x</code>. Then, we combine this data (as shown below) with the metadata to use different aesthetics and colors on the plot.</p>
<pre class="r"><code>vst_pca$x</code></pre>
<details>
<p><summary>output</summary></p>
<pre><code>##                  PC1        PC2        PC3         PC4       PC5         PC6
## Sample_1  -34.641623 -13.435531   6.486032  -0.2742477 -6.406502   5.6703279
## Sample_2  -35.431333 -13.277983   5.171377  -0.3540945 -3.953216   6.9974341
## Sample_3  -35.938795 -14.544994   2.885351 -11.3414829  5.950082 -10.8497916
## Sample_4  -20.672358   3.962013  -1.414301  10.4819194 -0.195882  -2.8003847
## Sample_5  -21.155503   1.390981  -6.644132   7.4002617  6.505502   0.9771040
## Sample_6  -22.662075   1.115504  -9.801356   6.3107519  7.431833   0.6592987
## Sample_7    1.862762  24.449057  12.865650  -4.4029501  5.822040   4.1873901
## Sample_8   -5.909698  13.992629 -14.775686 -10.9789567 -8.967097   2.2612207
## Sample_9   -3.233544  17.321871   5.196746   2.5142966 -6.983077  -7.2580550
## Sample_10  60.630406  -5.930071   6.993097   5.7393184 -4.254238  -2.6905179
## Sample_11  56.669696 -10.638879  -5.860598   0.4646609 -5.216763  -1.3748865
## Sample_12  60.482067  -4.404596  -1.102179  -5.5594769 10.267317   4.2208603
##                  PC7        PC8        PC9       PC10       PC11         PC12
## Sample_1  -7.6896364  7.5508580 -1.1936551 -0.2382806 -0.7901002 6.418477e-15
## Sample_2   5.8962070 -9.0662471 -1.0449651  0.2284545  1.3982703 1.526557e-15
## Sample_3   1.4353811  0.7461021  1.4769949  0.2114041 -0.5247079 4.045375e-15
## Sample_4   1.2992058  2.3863551  6.1089574 -0.1510797  9.1851796 3.588969e-14
## Sample_5  -2.7487548 -2.7353990  4.5686381 -6.6024251 -6.9911571 5.848273e-14
## Sample_6   2.1391551  2.3567136 -7.7925047  6.9400619 -1.3499685 4.436035e-14
## Sample_7   5.2166866  4.0805978  0.2292610 -1.4574679 -1.2200771 4.266032e-14
## Sample_8  -0.1684735 -0.1448735  3.9822133  2.5610601 -0.5182664 5.483808e-14
## Sample_9  -5.4663572 -5.2056738 -6.4295279 -1.6062665  0.5133496 4.941880e-14
## Sample_10  0.8277618 -0.5767198  4.9156730  6.8596253 -4.8748004 4.991840e-14
## Sample_11  5.8886122  3.9566331 -4.4374522 -6.9835771  0.4457707 1.114872e-13
## Sample_12 -6.6297877 -3.3483466 -0.3836326  0.2384911  4.7265074 9.376180e-14</code></pre>
</details>
<p>And, if you check the <code>class()</code> of this object, you will realize that this is a <code>matrix</code>. To be able to comfortably use <code>tidyverse</code> on this object, we must first convert this to a <code>data.frame</code>.</p>
<pre class="r"><code>vst_pca_all &lt;- vst_pca$x %&gt;%
  as.data.frame() %&gt;%
  rownames_to_column(var = &quot;Sample_ID&quot;) %&gt;%
  full_join(md, by = &quot;Sample_ID&quot;)

# Just to keep the order the right way.
vst_pca_all$Sample_Name &lt;- factor(vst_pca_all$Sample_Name, levels = c(&quot;t0_A&quot;,&quot;t0_B&quot;,&quot;t0_C&quot;,&quot;t2_A&quot;,&quot;t2_B&quot;,&quot;t2_C&quot;,&quot;t6_A&quot;,&quot;t6_B&quot;,&quot;t6_C&quot;,&quot;t24_A&quot;,&quot;t24_B&quot;,&quot;t24_C&quot;))
vst_pca_all$Time &lt;- factor(vst_pca_all$Time, levels = c(&quot;t0&quot;,&quot;t2&quot;,&quot;t6&quot;,&quot;t24&quot;))
vst_pca_all$Replicate &lt;- factor(vst_pca_all$Replicate, levels = c(&quot;A&quot;,&quot;B&quot;,&quot;C&quot;))

ggplot(vst_pca_all, aes(x=PC1, y=PC2, color = Time)) +
  geom_point(size = 3, aes(shape = Replicate)) +
  geom_vline(xintercept = 0, linetype=2) +
  geom_hline(yintercept = 0, linetype=2) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank()) </code></pre>
<details>
<p><summary>plot</summary></p>
<p><img src="lab_pca_hmap_files/figure-html/unnamed-chunk-7-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
</div>
<div id="loading-plot" class="section level2">
<h2><span class="header-section-number">1.3</span> Loading plot</h2>
<p>Now, let us say you want to plot the four genes that contribute the most to the four directions in the PCA plot. We could obtain them by looking at the <code>vst_pca$rotation</code> matrix. We could get those genes and their respective coordinates as follows.</p>
<pre class="r"><code>genes.selected=vst_pca$rotation[c(which.max(vst_pca$rotation[,&quot;PC1&quot;]), which.min(vst_pca$rotation[,&quot;PC1&quot;]), which.max(vst_pca$rotation[,&quot;PC2&quot;]), which.min(vst_pca$rotation[,&quot;PC2&quot;])),c(&quot;PC1&quot;,&quot;PC2&quot;)]
genes.selected &lt;- genes.selected %&gt;%
  as.data.frame() %&gt;%
  rownames_to_column(var = &quot;Gene_ID&quot;)
genes.selected</code></pre>
<details>
<p><summary>output</summary></p>
<pre><code>##           Gene_ID         PC1         PC2
## 1 ENSG00000204941  0.05737805 -0.02981714
## 2 ENSG00000175592 -0.04640560 -0.03986039
## 3 ENSG00000156804  0.01143268  0.03864008
## 4 ENSG00000120738  0.01010573 -0.05681563</code></pre>
</details>
<p>A loading plot shows how strongly each variable (gene) influences a principal component. As an example, we could plot the four genes we selected.</p>
<pre class="r"><code>ggplot(genes.selected, aes(x=PC1, y=PC2)) +
  geom_point() +
  geom_segment(aes(xend=PC1, yend=PC2), x=0, y=0, color=&quot;Grey&quot;) +
  geom_label(aes(x=PC1, y=PC2, label=Gene_ID), size=2, vjust=&quot;outward&quot;) +
  theme_bw() +
  theme(panel.grid.major = element_blank(),  panel.grid.minor = element_blank()) </code></pre>
<details>
<p><summary>plot</summary></p>
<p><img src="lab_pca_hmap_files/figure-html/unnamed-chunk-9-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
</div>
</div>
<div id="pca-bi-plot" class="section level1">
<h1><span class="header-section-number">2</span> PCA bi-plot</h1>
<p>By merging the PCA plot with the loadings plot one can create a so-called PCA bi-plot.</p>
<pre class="r"><code>scale=500
ggplot(data=vst_pca_all, mapping=aes(x=PC1, y=PC2)) +
     geom_point(size = 3, aes(shape = Replicate, color = Time)) +
     geom_vline(xintercept = 0, linetype=2) +
     geom_hline(yintercept = 0, linetype=2) +
     geom_segment(data=genes.selected, mapping=aes(xend=scale*PC1,yend=scale*PC2), x=0, y=0, arrow=arrow(), color=&quot;grey&quot;) +
     geom_label(data=genes.selected, mapping=aes(x=scale*PC1,y=scale*PC2, label=Gene_ID), size=2, hjust=&quot;outward&quot;, vjust=&quot;outward&quot;) +
     theme_bw() +
     theme(panel.grid.major = element_blank(),  panel.grid.minor = element_blank())</code></pre>
<details>
<p><summary>plot</summary></p>
<p><img src="lab_pca_hmap_files/figure-html/unnamed-chunk-10-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
<p><i class="fas fa-exclamation-circle"></i> Note   Similarly, let us say, you have environmental variables (continous variables like pH and so on) from the same samples and you would like to see how they would fit this bi-plot. One can use <code>envfit()</code> function from the <code>vegan</code> package, this function would return both the <code>p-value</code> and the <code>coordinates</code> of each of the variables in your environmental matrix. Then you could subset the significant variables and plot them in the same way as above.</p>
</div>
<div id="heatmap" class="section level1">
<h1><span class="header-section-number">3</span> Heatmap</h1>
<div id="pheatmap" class="section level2">
<h2><span class="header-section-number">3.1</span> pheatmap</h2>
<p>For heatmap, let us look into <code>pheatmap</code> library which is not part of <code>ggplot</code>, but it is a well known package for building heatmaps. It contains a lot of internal aesthetics that you can add that are very informative and intuitive. Let us first start with making a correlation matrix and plot it.</p>
<pre class="r"><code>vst_cor &lt;- as.matrix(cor(gc_vst, method=&quot;spearman&quot;))
library(pheatmap)

pheatmap(vst_cor,border_color=NA,annotation_col=md[,&quot;Time&quot;,drop=F],
         annotation_row=md[,&quot;Time&quot;,drop=F],annotation_legend=T)</code></pre>
<details>
<p><summary>plot</summary></p>
<p><img src="lab_pca_hmap_files/figure-html/unnamed-chunk-11-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
<p><i class="fas fa-exclamation-circle"></i> Note   If you get a <code>gpar()</code> error, this is due to the reason that the metadata table you provide here does not have the sample ids as <code>rownames</code>. <code>pheatmap</code> maps the meta-data to the count data using <code>rownames</code></p>
<p>Now, this is based on a correlation matrix where you have a very simple square matrix with samples and their correlations to each to each other. Now let us look into how we can make a heatmap from an expression dataset.</p>
<p>For this we will use the dataset: <code>Time_t2_vs_t0.txt</code> here that basically has the list of genes that are differentially expressed in <code>t2</code> vs <code>t0</code> and we would like to visualize this in an heatmap. To be more precise, these are the top and bottom 200 (2 X 100) genes with <code>adjusted p-value &lt; 0.01</code>. We use just these 200, for the sake of visualization. Let us take the <code>t24 vs t0</code> here as these are most different based on our PCA.</p>
<p><i class="fas fa-exclamation-circle"></i> Note   The genes are already sorted in the file. First 100 genes with <code>negative log2FoldChange</code> and the next 100 with <code>positive log2FoldChange</code></p>
<pre class="r"><code>diff_t2_vs_t0 &lt;- read.table(&quot;data/Time_t24_vs_t0.txt&quot;, sep = &quot;\t&quot;, header = TRUE, row.names = 1)
hmap_t2_t0 &lt;- subset(gc_vst, rownames(gc_vst) %in% diff_t2_vs_t0$gene)

pheatmap(hmap_t2_t0, border_color=NA, annotation_col=md[,&quot;Time&quot;,drop=F], annotation_row=NULL, annotation_legend=T)</code></pre>
<details>
<p><summary>plot</summary></p>
<p><img src="lab_pca_hmap_files/figure-html/unnamed-chunk-12-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
<p>Further, we can also customize using the <code>pheatmap</code> and also use the continous color options we talked about in the earlier sections.</p>
<pre class="r"><code>library(wesanderson)
md$Time &lt;- factor(md$Time, levels = c(&quot;t0&quot;,&quot;t2&quot;,&quot;t6&quot;,&quot;t24&quot;))

pheatmap(hmap_t2_t0, color = wes_palette(&quot;Moonrise3&quot;, 100, type = &quot;continuous&quot;), border_color=NA, annotation_col=md[,&quot;Time&quot;,drop=F], annotation_row=NULL, annotation_legend=T, show_rownames = FALSE, fontsize = 14)</code></pre>
<details>
<p><summary>plot</summary></p>
<p><img src="lab_pca_hmap_files/figure-html/unnamed-chunk-13-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
</div>
<div id="ggplot" class="section level2">
<h2><span class="header-section-number">3.2</span> ggplot</h2>
<p>We can use <code>geom_tile()</code> from <code>ggplot</code> to make heatmaps. The <code>ggplot</code> based heatmaps are, as we have seen for other cases, much more easier to customize.</p>
<pre class="r"><code>hmap_t2_t0_long &lt;- hmap_t2_t0 %&gt;%
  rownames_to_column(var = &quot;Gene&quot;) %&gt;%
  gather(Sample_ID, VST, -Gene) %&gt;% 
  full_join(md, by = &quot;Sample_ID&quot;)  

hmap_t2_t0_long$Sample_Name &lt;- factor(hmap_t2_t0_long$Sample_Name, levels = 
                                        c(&quot;t0_A&quot;,&quot;t0_B&quot;,&quot;t0_C&quot;,&quot;t2_A&quot;,&quot;t2_B&quot;,&quot;t2_C&quot;,&quot;t6_A&quot;,&quot;t6_B&quot;,&quot;t6_C&quot;,&quot;t24_A&quot;,&quot;t24_B&quot;,&quot;t24_C&quot;))
hmap_t2_t0_long$Time &lt;- factor(hmap_t2_t0_long$Time, levels = c(&quot;t0&quot;,&quot;t2&quot;,&quot;t6&quot;,&quot;t24&quot;))
hmap_t2_t0_long$Replicate &lt;- factor(hmap_t2_t0_long$Replicate, levels = c(&quot;A&quot;,&quot;B&quot;,&quot;C&quot;))
hmap_t2_t0_long$Gene &lt;- factor(hmap_t2_t0_long$Gene, levels = row.names(hmap_t2_t0))

ggplot(hmap_t2_t0_long) +
  geom_tile(aes(x = Sample_Name, y = Gene, fill = VST)) +
  scale_fill_gradientn(colours = rainbow(5)) +
  scale_x_discrete(limits = c(&quot;t0_A&quot;,&quot;t0_B&quot;,&quot;t0_C&quot;,&quot;t24_A&quot;,&quot;t24_B&quot;,&quot;t24_C&quot;)) +
  theme(axis.text.y = element_blank(), axis.ticks = element_blank()) </code></pre>
<details>
<p><summary>plot</summary></p>
<p><img src="lab_pca_hmap_files/figure-html/unnamed-chunk-14-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
</div>
</div>
<div id="exercise" class="section level1">
<h1><span class="header-section-number">4</span> Exercise</h1>
<p>Now, as I have mentioned earlier, building a plot similar to PCA really depends on how the object looks like. Now, let us try to make a <code>MDS</code> or <code>PCoA</code> plot from the same data as we have used. Here is how you get the MDS object in R.</p>
<pre class="r"><code>gc_dist &lt;- dist(t(gc_vst))
gc_mds &lt;- cmdscale(gc_dist,eig=TRUE, k=2) </code></pre>
<p><i class="fas fa-clipboard-list"></i> Task   Now, try to replicate the example <code>MDS</code> plot below if you have enough time.</p>
<details>
<p><summary>plot</summary></p>
<p><img src="lab_pca_hmap_files/figure-html/unnamed-chunk-16-1.svg" width="672" style="display: block; margin: auto auto auto 0;" /></p>
</details>
<p><i class="fas fa-clipboard-list"></i> Hint   The mds object <code>gc_mds</code> has “eigenvalues”. You can calculate the variance by <code>Variance &lt;- Eigenvalues / sum(Eigenvalues)</code></p>
</div>
<div id="session-info" class="section level1">
<h1><span class="header-section-number">5</span> Session info</h1>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 4.1.3 (2022-03-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.6 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
## LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so
## 
## locale:
##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] pheatmap_1.0.12        swemaps_1.0            mapdata_2.3.0         
##  [4] maps_3.4.0             ggpubr_0.4.0           cowplot_1.1.1         
##  [7] ggthemes_4.2.4         scales_1.2.0           ggrepel_0.9.1         
## [10] wesanderson_0.3.6      forcats_0.5.1          stringr_1.4.0         
## [13] purrr_0.3.4            readr_2.1.2            tidyr_1.2.0           
## [16] tibble_3.1.6           tidyverse_1.3.1        reshape2_1.4.4        
## [19] ggplot2_3.3.5          formattable_0.2.1      kableExtra_1.3.4      
## [22] dplyr_1.0.8            lubridate_1.8.0        leaflet_2.1.1         
## [25] yaml_2.3.5             fontawesome_0.2.2.9000 captioner_2.2.3       
## [28] bookdown_0.26          knitr_1.38            
## 
## loaded via a namespace (and not attached):
##  [1] nlme_3.1-155            fs_1.5.2                RColorBrewer_1.1-3     
##  [4] webshot_0.5.3           httr_1.4.2              tools_4.1.3            
##  [7] backports_1.4.1         bslib_0.3.1             utf8_1.2.2             
## [10] R6_2.5.1                DBI_1.1.2               mgcv_1.8-39            
## [13] colorspace_2.0-3        withr_2.5.0             gridExtra_2.3          
## [16] tidyselect_1.1.2        compiler_4.1.3          cli_3.3.0              
## [19] rvest_1.0.2             xml2_1.3.3              labeling_0.4.2         
## [22] sass_0.4.1              systemfonts_1.0.4       digest_0.6.29          
## [25] rmarkdown_2.14          svglite_2.1.0           pkgconfig_2.0.3        
## [28] htmltools_0.5.2         dbplyr_2.1.1            fastmap_1.1.0          
## [31] highr_0.9               htmlwidgets_1.5.4       rlang_1.0.2            
## [34] readxl_1.4.0            rstudioapi_0.13         jquerylib_0.1.4        
## [37] farver_2.1.0            generics_0.1.2          jsonlite_1.8.0         
## [40] crosstalk_1.2.0         car_3.0-12              magrittr_2.0.3         
## [43] Matrix_1.4-0            Rcpp_1.0.8.3            munsell_0.5.0          
## [46] fansi_1.0.3             abind_1.4-5             lifecycle_1.0.1        
## [49] stringi_1.7.6           carData_3.0-5           plyr_1.8.7             
## [52] grid_4.1.3              crayon_1.5.1            lattice_0.20-45        
## [55] haven_2.5.0             splines_4.1.3           hms_1.1.1              
## [58] pillar_1.7.0            ggsignif_0.6.3          reprex_2.0.1           
## [61] glue_1.6.2              evaluate_0.15           leaflet.providers_1.9.0
## [64] modelr_0.1.8            vctrs_0.4.1             tzdb_0.3.0             
## [67] cellranger_1.1.0        gtable_0.3.0            assertthat_0.2.1       
## [70] xfun_0.30               broom_0.8.0             rstatix_0.7.0          
## [73] viridisLite_0.4.0       ellipsis_0.3.2</code></pre>
<hr />
</div>

<footer class="footer">
<div class="container container-footer">
<div class="row">
<div class="col-sm-8" style="text-align:left;vertical-align:middle;">
<p style="color:#bdbdbd;">
<b>2022</b> • NBIS • SciLifeLab
</p>
</div>
<div class="col-sm-4" style="text-align:right;vertical-align:middle;">
<p style="color:#bdbdbd;">
<span> <span class="footericon" style="padding-right:4px; padding-left:4px"> <a href="https://nbis.se/"><svg aria-hidden="true" role="img" viewBox="0 0 496 512" style="height:1em;width:0.97em;vertical-align:-0.125em;margin-left:auto;margin-right:auto;font-size:inherit;fill:#bdbdbd;overflow:visible;position:relative;"><path d="M248 8C111.03 8 0 119.03 0 256s111.03 248 248 248 248-111.03 248-248S384.97 8 248 8zm82.29 357.6c-3.9 3.88-7.99 7.95-11.31 11.28-2.99 3-5.1 6.7-6.17 10.71-1.51 5.66-2.73 11.38-4.77 16.87l-17.39 46.85c-13.76 3-28 4.69-42.65 4.69v-27.38c1.69-12.62-7.64-36.26-22.63-51.25-6-6-9.37-14.14-9.37-22.63v-32.01c0-11.64-6.27-22.34-16.46-27.97-14.37-7.95-34.81-19.06-48.81-26.11-11.48-5.78-22.1-13.14-31.65-21.75l-.8-.72a114.792 114.792 0 0 1-18.06-20.74c-9.38-13.77-24.66-36.42-34.59-51.14 20.47-45.5 57.36-82.04 103.2-101.89l24.01 12.01C203.48 89.74 216 82.01 216 70.11v-11.3c7.99-1.29 16.12-2.11 24.39-2.42l28.3 28.3c6.25 6.25 6.25 16.38 0 22.63L264 112l-10.34 10.34c-3.12 3.12-3.12 8.19 0 11.31l4.69 4.69c3.12 3.12 3.12 8.19 0 11.31l-8 8a8.008 8.008 0 0 1-5.66 2.34h-8.99c-2.08 0-4.08.81-5.58 2.27l-9.92 9.65a8.008 8.008 0 0 0-1.58 9.31l15.59 31.19c2.66 5.32-1.21 11.58-7.15 11.58h-5.64c-1.93 0-3.79-.7-5.24-1.96l-9.28-8.06a16.017 16.017 0 0 0-15.55-3.1l-31.17 10.39a11.95 11.95 0 0 0-8.17 11.34c0 4.53 2.56 8.66 6.61 10.69l11.08 5.54c9.41 4.71 19.79 7.16 30.31 7.16s22.59 27.29 32 32h66.75c8.49 0 16.62 3.37 22.63 9.37l13.69 13.69a30.503 30.503 0 0 1 8.93 21.57 46.536 46.536 0 0 1-13.72 32.98zM417 274.25c-5.79-1.45-10.84-5-14.15-9.97l-17.98-26.97a23.97 23.97 0 0 1 0-26.62l19.59-29.38c2.32-3.47 5.5-6.29 9.24-8.15l12.98-6.49C440.2 193.59 448 223.87 448 256c0 8.67-.74 17.16-1.82 25.54L417 274.25z"></path></svg></a> </span> <span class="footericon" style="padding-right:4px; padding-left:4px"> <a href="https://twitter.com/NBISwe"><svg aria-hidden="true" role="img" viewBox="0 0 512 512" style="height:1em;width:1em;vertical-align:-0.125em;margin-left:auto;margin-right:auto;font-size:inherit;fill:#bdbdbd;overflow:visible;position:relative;"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"></path></svg></a> </span> <span class="footericon" style="padding-left:4px"> <a href="https://www.linkedin.com/company/nbisweden/"><svg aria-hidden="true" role="img" viewBox="0 0 448 512" style="height:1em;width:0.88em;vertical-align:-0.125em;margin-left:auto;margin-right:auto;font-size:inherit;fill:#bdbdbd;overflow:visible;position:relative;"><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"></path></svg></a> </span> </span>
</p>
</div>
</div>
</div>
</footer>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
