<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Lokesh Mano • 11-Oct-2022" />


<title>Intro to R</title>

<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<link id="font-awesome-1-attachment" rel="attachment" href="site_libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
#TOC {
  display: none !important;
}
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div id="header">





<h4 class="author"><b>Lokesh Mano</b> • 11-Oct-2022</h4>

</div>


<!-- rmd lab header -->
<p><br></p>
<p>R is a programming language for statistical computing, and data wrangling. It is open-source, widely used in data science, has a wide range of functions and algorithms for graphing and data analyses.</p>
<p>We will use R in this course for generating plots from different biological data that are of higher quality and standard for publications. To do this, you will brush-up your memory on some important aspects of R that are important for this course below:</p>
<p><strong>Before starting with the lab sessions of the entire course, you must have downloaded all the necessary files from [link removed] and make sure that the directory tree looks similar to the one displayed in that page</strong></p>
<p>In that case, you can proceed with the exercise now and remember to have fun :)</p>
<div id="inputoutput" class="section level1">
<h1><span class="header-section-number">1</span> Input/Output</h1>
<p>Input and output of data and images is an important aspect with data analysis.</p>
<div id="text" class="section level2">
<h2><span class="header-section-number">1.1</span> Text</h2>
<p>Data can come in a variety of formats which needs to be read into R and converted to an R data type.</p>
<p>Text files are the most commonly used input. Text files can be read in using the function <code>read.table</code>. We have a sample file to use: <strong>iris.txt</strong>.</p>
<pre class="r"><code>dfr &lt;- read.table(&quot;data/metadata.csv&quot;,sep = &quot;;&quot;, header=TRUE,stringsAsFactors=F)</code></pre>
<p>This reads in a tab-delimited text file with a header. The argument <code>sep='\t'</code> is set by default to specify that the delimiter is a tab. <code>stringsAsFactors=F</code> setting ensures that character columns are not automatically converted to factors.</p>
<p>It’s always a good idea to check the data after import.</p>
<pre class="r"><code>head(dfr)</code></pre>
<details>
<p><summary>output</summary></p>
<pre><code>##   Sample_ID Sample_Name Time Replicate Cell
## 1  Sample_1        t0_A   t0         A A431
## 2  Sample_2        t0_B   t0         B A431
## 3  Sample_3        t0_C   t0         C A431
## 4  Sample_4        t2_A   t2         A A431
## 5  Sample_5        t2_B   t2         B A431
## 6  Sample_6        t2_C   t2         C A431</code></pre>
</details>
<pre class="r"><code>str(dfr)</code></pre>
<details>
<p><summary>output</summary></p>
<pre><code>## &#39;data.frame&#39;:    12 obs. of  5 variables:
##  $ Sample_ID  : chr  &quot;Sample_1&quot; &quot;Sample_2&quot; &quot;Sample_3&quot; &quot;Sample_4&quot; ...
##  $ Sample_Name: chr  &quot;t0_A&quot; &quot;t0_B&quot; &quot;t0_C&quot; &quot;t2_A&quot; ...
##  $ Time       : chr  &quot;t0&quot; &quot;t0&quot; &quot;t0&quot; &quot;t2&quot; ...
##  $ Replicate  : chr  &quot;A&quot; &quot;B&quot; &quot;C&quot; &quot;A&quot; ...
##  $ Cell       : chr  &quot;A431&quot; &quot;A431&quot; &quot;A431&quot; &quot;A431&quot; ...</code></pre>
</details>
<p>Check <code>?read.table</code> for other wrapper functions to read in text files.</p>
<p>Let’s filter this data.frame and create a new data set.</p>
<pre class="r"><code>dfr1 &lt;- dfr[dfr$Time == &quot;t0&quot;,]</code></pre>
<p>And we can write this as a text file.</p>
<pre class="r"><code>write.table(dfr1,&quot;iris-setosa.txt&quot;,sep=&quot;\t&quot;,row.names=F,quote=F)</code></pre>
<p><code>sep="\t"</code> sets the delimiter to tab. <code>row.names=F</code> denotes that rownames should not be written. <code>quote=F</code> specifies that doubles must not be placed around strings.</p>
</div>
</div>
<div id="data-frames" class="section level1">
<h1><span class="header-section-number">2</span> Data-Frames</h1>
<p>You have probably learnt about <code>data.frame</code> in your previous course and this is the most important data structure for generating plots using <code>ggplot</code>. So, below you have some quick exercises on how to work with Data-Frames.</p>
<p>Vectors positions can be accessed using <code>[]</code>. R follows 1-based indexing, meaning that the indexing starts at 1.</p>
<p>Data-frame or matrix positions can be accessed using <code>[]</code> specifying row and column like <code>[row,column]</code>.</p>
<pre class="r"><code>dfr &lt;- data.frame(x = 1:3, y = c(&quot;a&quot;, &quot;b&quot;, &quot;c&quot;))
dfr
dfr[1,]
dfr[,1]
dfr[2,2]</code></pre>
<details>
<p><summary>output</summary></p>
<pre><code>##   x y
## 1 1 a
## 2 2 b
## 3 3 c
##   x y
## 1 1 a
## [1] 1 2 3
## [1] &quot;b&quot;</code></pre>
</details>
<p>The function <code>cbind()</code> is used to join two data-frames column-wise.</p>
<pre class="r"><code>dfr1 &lt;- data.frame(x = 1:3, y = c(&quot;a&quot;, &quot;b&quot;, &quot;c&quot;))
dfr2 &lt;- data.frame(p = 4:6, q = c(&quot;d&quot;, &quot;e&quot;, &quot;f&quot;))
dfr1
dfr2

cbind(dfr1,dfr2)</code></pre>
<details>
<p><summary>output</summary></p>
<pre><code>##   x y
## 1 1 a
## 2 2 b
## 3 3 c
##   p q
## 1 4 d
## 2 5 e
## 3 6 f
##   x y p q
## 1 1 a 4 d
## 2 2 b 5 e
## 3 3 c 6 f</code></pre>
</details>
<p>Similarly, <code>rbind()</code> is used to join two data-frames row-wise.</p>
<pre class="r"><code>dfr1 &lt;- data.frame(x = 1:3, y = c(&quot;a&quot;, &quot;b&quot;, &quot;c&quot;))
dfr2 &lt;- data.frame(x = 4:6, y = c(&quot;d&quot;, &quot;e&quot;, &quot;f&quot;))
dfr1
dfr2

rbind(dfr1,dfr2)</code></pre>
<details>
<p><summary>output</summary></p>
<pre><code>##   x y
## 1 1 a
## 2 2 b
## 3 3 c
##   x y
## 1 4 d
## 2 5 e
## 3 6 f
##   x y
## 1 1 a
## 2 2 b
## 3 3 c
## 4 4 d
## 5 5 e
## 6 6 f</code></pre>
</details>
<p>Two data-frames can be merged based on a shared column using the <code>merge()</code> function.</p>
<pre class="r"><code>dfr1 &lt;- data.frame(x = 1:4, p = c(&quot;a&quot;, &quot;b&quot;, &quot;c&quot;,&quot;d&quot;))
dfr2 &lt;- data.frame(x = 3:6, q = c(&quot;l&quot;, &quot;m&quot;, &quot;n&quot;,&quot;o&quot;))
dfr1
dfr2

merge(dfr1,dfr2,by=&quot;x&quot;)
merge(dfr1,dfr2,by=&quot;x&quot;,all.x=T)
merge(dfr1,dfr2,by=&quot;x&quot;,all.y=T)
merge(dfr1,dfr2,by=&quot;x&quot;,all=T)</code></pre>
<details>
<p><summary>output</summary></p>
<pre><code>##   x p
## 1 1 a
## 2 2 b
## 3 3 c
## 4 4 d
##   x q
## 1 3 l
## 2 4 m
## 3 5 n
## 4 6 o
##   x p q
## 1 3 c l
## 2 4 d m
##   x p    q
## 1 1 a &lt;NA&gt;
## 2 2 b &lt;NA&gt;
## 3 3 c    l
## 4 4 d    m
##   x    p q
## 1 3    c l
## 2 4    d m
## 3 5 &lt;NA&gt; n
## 4 6 &lt;NA&gt; o
##   x    p    q
## 1 1    a &lt;NA&gt;
## 2 2    b &lt;NA&gt;
## 3 3    c    l
## 4 4    d    m
## 5 5 &lt;NA&gt;    n
## 6 6 &lt;NA&gt;    o</code></pre>
</details>
</div>
<div id="data-formats" class="section level1">
<h1><span class="header-section-number">3</span> Data formats</h1>
<p>In terms of the R (and other similar programming languages), the data can be viewed or stored in two main formats! They are called <code>wide</code> and <code>long</code> formats. Below you will see what exactly they stand for and why is it important for plotting in ggplot.</p>
<div id="wide-format" class="section level2">
<h2><span class="header-section-number">3.1</span> Wide format</h2>
<p>A quick preview:</p>
<p><strong>Counts Table</strong></p>
<table class="table table-striped" style="width: auto !important; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
Sample_1
</th>
<th style="text-align:right;">
Sample_2
</th>
<th style="text-align:right;">
Sample_3
</th>
<th style="text-align:right;">
Sample_4
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
ENSG00000000003
</td>
<td style="text-align:right;">
321
</td>
<td style="text-align:right;">
303
</td>
<td style="text-align:right;">
204
</td>
<td style="text-align:right;">
492
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSG00000000005
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSG00000000419
</td>
<td style="text-align:right;">
696
</td>
<td style="text-align:right;">
660
</td>
<td style="text-align:right;">
472
</td>
<td style="text-align:right;">
951
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSG00000000457
</td>
<td style="text-align:right;">
59
</td>
<td style="text-align:right;">
54
</td>
<td style="text-align:right;">
44
</td>
<td style="text-align:right;">
109
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSG00000000460
</td>
<td style="text-align:right;">
399
</td>
<td style="text-align:right;">
405
</td>
<td style="text-align:right;">
236
</td>
<td style="text-align:right;">
445
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSG00000000938
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
</tbody>
</table>
<p>And we usually have our metadata related to the samples in another table like below:</p>
<p><strong>Metadata Table</strong></p>
<table class="table table-striped" style="width: auto !important; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
Sample_ID
</th>
<th style="text-align:left;">
Sample_Name
</th>
<th style="text-align:left;">
Time
</th>
<th style="text-align:left;">
Replicate
</th>
<th style="text-align:left;">
Cell
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Sample_1
</td>
<td style="text-align:left;">
t0_A
</td>
<td style="text-align:left;">
t0
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
A431
</td>
</tr>
<tr>
<td style="text-align:left;">
Sample_2
</td>
<td style="text-align:left;">
t0_B
</td>
<td style="text-align:left;">
t0
</td>
<td style="text-align:left;">
B
</td>
<td style="text-align:left;">
A431
</td>
</tr>
<tr>
<td style="text-align:left;">
Sample_3
</td>
<td style="text-align:left;">
t0_C
</td>
<td style="text-align:left;">
t0
</td>
<td style="text-align:left;">
C
</td>
<td style="text-align:left;">
A431
</td>
</tr>
<tr>
<td style="text-align:left;">
Sample_4
</td>
<td style="text-align:left;">
t2_A
</td>
<td style="text-align:left;">
t2
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
A431
</td>
</tr>
</tbody>
</table>
<ul>
<li>Wide format data is called “wide” because it typically has a lot of columns which stretch widely across the page or your computer screen.</li>
<li>Most of us are familiar with looking at wide format data
<ul>
<li>It is convenient and we are more used to looking at data this way in our Excel sheets.</li>
<li>It often lets you see more of the data, at one time, on your screen</li>
</ul></li>
</ul>
</div>
<div id="long-format" class="section level2">
<h2><span class="header-section-number">3.2</span> Long format</h2>
<p>Below is glimpse how the long format of the same data look like:</p>
<table class="table table-striped" style="width: auto !important; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
Gene
</th>
<th style="text-align:left;">
Samples
</th>
<th style="text-align:right;">
count
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
ENSG00000000003
</td>
<td style="text-align:left;">
Sample_1
</td>
<td style="text-align:right;">
321
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSG00000000005
</td>
<td style="text-align:left;">
Sample_1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSG00000000419
</td>
<td style="text-align:left;">
Sample_1
</td>
<td style="text-align:right;">
696
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSG00000000457
</td>
<td style="text-align:left;">
Sample_1
</td>
<td style="text-align:right;">
59
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSG00000000460
</td>
<td style="text-align:left;">
Sample_1
</td>
<td style="text-align:right;">
399
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSG00000000938
</td>
<td style="text-align:left;">
Sample_1
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSG00000000003
</td>
<td style="text-align:left;">
Sample_2
</td>
<td style="text-align:right;">
303
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSG00000000005
</td>
<td style="text-align:left;">
Sample_2
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSG00000000419
</td>
<td style="text-align:left;">
Sample_2
</td>
<td style="text-align:right;">
660
</td>
</tr>
<tr>
<td style="text-align:left;">
ENSG00000000457
</td>
<td style="text-align:left;">
Sample_2
</td>
<td style="text-align:right;">
54
</td>
</tr>
</tbody>
</table>
<p>Or to be even more complete and precise:</p>
<table class="table table-striped" style="width: auto !important; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
Sample_ID
</th>
<th style="text-align:left;">
Sample_Name
</th>
<th style="text-align:left;">
Time
</th>
<th style="text-align:left;">
Replicate
</th>
<th style="text-align:left;">
Cell
</th>
<th style="text-align:left;">
Gene
</th>
<th style="text-align:right;">
count
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Sample_1
</td>
<td style="text-align:left;">
t0_A
</td>
<td style="text-align:left;">
t0
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
A431
</td>
<td style="text-align:left;">
ENSG00000000003
</td>
<td style="text-align:right;">
321
</td>
</tr>
<tr>
<td style="text-align:left;">
Sample_1
</td>
<td style="text-align:left;">
t0_A
</td>
<td style="text-align:left;">
t0
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
A431
</td>
<td style="text-align:left;">
ENSG00000000005
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Sample_1
</td>
<td style="text-align:left;">
t0_A
</td>
<td style="text-align:left;">
t0
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
A431
</td>
<td style="text-align:left;">
ENSG00000000419
</td>
<td style="text-align:right;">
696
</td>
</tr>
<tr>
<td style="text-align:left;">
Sample_1
</td>
<td style="text-align:left;">
t0_A
</td>
<td style="text-align:left;">
t0
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
A431
</td>
<td style="text-align:left;">
ENSG00000000457
</td>
<td style="text-align:right;">
59
</td>
</tr>
<tr>
<td style="text-align:left;">
Sample_1
</td>
<td style="text-align:left;">
t0_A
</td>
<td style="text-align:left;">
t0
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
A431
</td>
<td style="text-align:left;">
ENSG00000000460
</td>
<td style="text-align:right;">
399
</td>
</tr>
<tr>
<td style="text-align:left;">
Sample_1
</td>
<td style="text-align:left;">
t0_A
</td>
<td style="text-align:left;">
t0
</td>
<td style="text-align:left;">
A
</td>
<td style="text-align:left;">
A431
</td>
<td style="text-align:left;">
ENSG00000000938
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Sample_2
</td>
<td style="text-align:left;">
t0_B
</td>
<td style="text-align:left;">
t0
</td>
<td style="text-align:left;">
B
</td>
<td style="text-align:left;">
A431
</td>
<td style="text-align:left;">
ENSG00000000003
</td>
<td style="text-align:right;">
303
</td>
</tr>
<tr>
<td style="text-align:left;">
Sample_2
</td>
<td style="text-align:left;">
t0_B
</td>
<td style="text-align:left;">
t0
</td>
<td style="text-align:left;">
B
</td>
<td style="text-align:left;">
A431
</td>
<td style="text-align:left;">
ENSG00000000005
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Sample_2
</td>
<td style="text-align:left;">
t0_B
</td>
<td style="text-align:left;">
t0
</td>
<td style="text-align:left;">
B
</td>
<td style="text-align:left;">
A431
</td>
<td style="text-align:left;">
ENSG00000000419
</td>
<td style="text-align:right;">
660
</td>
</tr>
<tr>
<td style="text-align:left;">
Sample_2
</td>
<td style="text-align:left;">
t0_B
</td>
<td style="text-align:left;">
t0
</td>
<td style="text-align:left;">
B
</td>
<td style="text-align:left;">
A431
</td>
<td style="text-align:left;">
ENSG00000000457
</td>
<td style="text-align:right;">
54
</td>
</tr>
</tbody>
</table>
<ul>
<li>Long format data is typically less familiar to most humans
<ul>
<li>It seems awfully hard to get a good look at all (or most) of it</li>
<li>It seems like it would require more storage on your hard disk</li>
<li>It seems like it would be harder to enter data in a long format</li>
</ul></li>
</ul>
</div>
<div id="which-is-better" class="section level2">
<h2><span class="header-section-number">3.3</span> Which is better?</h2>
<ul>
<li>Well, there are some contexts where putting things in wide format is computationally efficient because you can treat data in a matrix format and to efficient matrix calculations on it.</li>
<li>However, adding data to wide format data sets is very hard:
<ol style="list-style-type: decimal">
<li>It is very difficult to conceive of analytic schemes that apply generally across all wide-format data sets.</li>
<li>Many tools in R want data in long format <strong>like ggplot</strong></li>
<li>The long format for data corresponds to the relational model for storing data, which is the model used in most modern data bases like the SQL family of data base systems.</li>
</ol></li>
<li>A more technical treatment of wide versus long data requires some terminology:
<ul>
<li><span style="color:blue">Identifier variables</span> are often categorical things that cross-classify observations into categories.</li>
<li><span style="color:red">Measured variables</span> are the names given to properties or characteristics that you can go out and measure.</li>
<li><span style="color:orange">Values</span> are the values that you measure are record for any particular measured variable.</li>
</ul></li>
<li>In any particular data set, what you might want to call an <span style="color:blue">Identifier variables</span> versus a <span style="color:red">Measured variables</span> can not always be entirely clear.
<ul>
<li>Other people might choose to define things differently.</li>
</ul></li>
<li>However, to my mind, it is less important to be able to precisely recognize these three entities in every possible situation (where there might be some fuzziness about which is which)</li>
<li>And it is more important to understand how <span style="color:blue">Identifier variables</span>, <span style="color:red">Measured variables</span>, and <span style="color:orange">Values</span> interact and behave when we are transforming data between wide and long formats.</li>
</ul>
</div>
</div>
<div id="conversion-between-formats" class="section level1">
<h1><span class="header-section-number">4</span> Conversion between formats</h1>
<p>As for the biological data analysis, to be able to use tools such as <strong>ggplot</strong>, in simple terms we should learn to convert our data:</p>
<p>As per our previous examples: We should learn to convert</p>
<p><strong>From this format</strong></p>
<table class="table table-striped" style="width: auto !important; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;font-weight: bold;color: blue !important;">
</th>
<th style="text-align:right;font-weight: bold;color: blue !important;">
Sample_1
</th>
<th style="text-align:right;font-weight: bold;color: blue !important;">
Sample_2
</th>
<th style="text-align:right;font-weight: bold;color: blue !important;">
Sample_3
</th>
<th style="text-align:right;font-weight: bold;color: blue !important;">
Sample_4
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;color: orange !important;color: red !important;">
ENSG00000000003
</td>
<td style="text-align:right;color: orange !important;">
321
</td>
<td style="text-align:right;color: orange !important;">
303
</td>
<td style="text-align:right;color: orange !important;">
204
</td>
<td style="text-align:right;color: orange !important;">
492
</td>
</tr>
<tr>
<td style="text-align:left;color: orange !important;color: red !important;">
ENSG00000000005
</td>
<td style="text-align:right;color: orange !important;">
0
</td>
<td style="text-align:right;color: orange !important;">
0
</td>
<td style="text-align:right;color: orange !important;">
0
</td>
<td style="text-align:right;color: orange !important;">
0
</td>
</tr>
<tr>
<td style="text-align:left;color: orange !important;color: red !important;">
ENSG00000000419
</td>
<td style="text-align:right;color: orange !important;">
696
</td>
<td style="text-align:right;color: orange !important;">
660
</td>
<td style="text-align:right;color: orange !important;">
472
</td>
<td style="text-align:right;color: orange !important;">
951
</td>
</tr>
<tr>
<td style="text-align:left;color: orange !important;color: red !important;">
ENSG00000000457
</td>
<td style="text-align:right;color: orange !important;">
59
</td>
<td style="text-align:right;color: orange !important;">
54
</td>
<td style="text-align:right;color: orange !important;">
44
</td>
<td style="text-align:right;color: orange !important;">
109
</td>
</tr>
<tr>
<td style="text-align:left;color: orange !important;color: red !important;">
ENSG00000000460
</td>
<td style="text-align:right;color: orange !important;">
399
</td>
<td style="text-align:right;color: orange !important;">
405
</td>
<td style="text-align:right;color: orange !important;">
236
</td>
<td style="text-align:right;color: orange !important;">
445
</td>
</tr>
<tr>
<td style="text-align:left;color: orange !important;color: red !important;">
ENSG00000000938
</td>
<td style="text-align:right;color: orange !important;">
0
</td>
<td style="text-align:right;color: orange !important;">
0
</td>
<td style="text-align:right;color: orange !important;">
0
</td>
<td style="text-align:right;color: orange !important;">
0
</td>
</tr>
</tbody>
</table>
<p><strong>To this format</strong></p>
<table class="table table-striped" style="width: auto !important; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
Sample_ID
</th>
<th style="text-align:left;">
Sample_Name
</th>
<th style="text-align:left;">
Time
</th>
<th style="text-align:left;">
Replicate
</th>
<th style="text-align:left;">
Cell
</th>
<th style="text-align:left;">
Gene
</th>
<th style="text-align:right;">
count
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;color: blue !important;">
Sample_1
</td>
<td style="text-align:left;color: blue !important;">
t0_A
</td>
<td style="text-align:left;color: blue !important;">
t0
</td>
<td style="text-align:left;color: blue !important;">
A
</td>
<td style="text-align:left;color: blue !important;">
A431
</td>
<td style="text-align:left;color: red !important;">
ENSG00000000003
</td>
<td style="text-align:right;color: orange !important;">
321
</td>
</tr>
<tr>
<td style="text-align:left;color: blue !important;">
Sample_1
</td>
<td style="text-align:left;color: blue !important;">
t0_A
</td>
<td style="text-align:left;color: blue !important;">
t0
</td>
<td style="text-align:left;color: blue !important;">
A
</td>
<td style="text-align:left;color: blue !important;">
A431
</td>
<td style="text-align:left;color: red !important;">
ENSG00000000005
</td>
<td style="text-align:right;color: orange !important;">
0
</td>
</tr>
<tr>
<td style="text-align:left;color: blue !important;">
Sample_1
</td>
<td style="text-align:left;color: blue !important;">
t0_A
</td>
<td style="text-align:left;color: blue !important;">
t0
</td>
<td style="text-align:left;color: blue !important;">
A
</td>
<td style="text-align:left;color: blue !important;">
A431
</td>
<td style="text-align:left;color: red !important;">
ENSG00000000419
</td>
<td style="text-align:right;color: orange !important;">
696
</td>
</tr>
<tr>
<td style="text-align:left;color: blue !important;">
Sample_1
</td>
<td style="text-align:left;color: blue !important;">
t0_A
</td>
<td style="text-align:left;color: blue !important;">
t0
</td>
<td style="text-align:left;color: blue !important;">
A
</td>
<td style="text-align:left;color: blue !important;">
A431
</td>
<td style="text-align:left;color: red !important;">
ENSG00000000457
</td>
<td style="text-align:right;color: orange !important;">
59
</td>
</tr>
<tr>
<td style="text-align:left;color: blue !important;">
Sample_1
</td>
<td style="text-align:left;color: blue !important;">
t0_A
</td>
<td style="text-align:left;color: blue !important;">
t0
</td>
<td style="text-align:left;color: blue !important;">
A
</td>
<td style="text-align:left;color: blue !important;">
A431
</td>
<td style="text-align:left;color: red !important;">
ENSG00000000460
</td>
<td style="text-align:right;color: orange !important;">
399
</td>
</tr>
<tr>
<td style="text-align:left;color: blue !important;">
Sample_1
</td>
<td style="text-align:left;color: blue !important;">
t0_A
</td>
<td style="text-align:left;color: blue !important;">
t0
</td>
<td style="text-align:left;color: blue !important;">
A
</td>
<td style="text-align:left;color: blue !important;">
A431
</td>
<td style="text-align:left;color: red !important;">
ENSG00000000938
</td>
<td style="text-align:right;color: orange !important;">
0
</td>
</tr>
<tr>
<td style="text-align:left;color: blue !important;">
Sample_2
</td>
<td style="text-align:left;color: blue !important;">
t0_B
</td>
<td style="text-align:left;color: blue !important;">
t0
</td>
<td style="text-align:left;color: blue !important;">
B
</td>
<td style="text-align:left;color: blue !important;">
A431
</td>
<td style="text-align:left;color: red !important;">
ENSG00000000003
</td>
<td style="text-align:right;color: orange !important;">
303
</td>
</tr>
<tr>
<td style="text-align:left;color: blue !important;">
Sample_2
</td>
<td style="text-align:left;color: blue !important;">
t0_B
</td>
<td style="text-align:left;color: blue !important;">
t0
</td>
<td style="text-align:left;color: blue !important;">
B
</td>
<td style="text-align:left;color: blue !important;">
A431
</td>
<td style="text-align:left;color: red !important;">
ENSG00000000005
</td>
<td style="text-align:right;color: orange !important;">
0
</td>
</tr>
<tr>
<td style="text-align:left;color: blue !important;">
Sample_2
</td>
<td style="text-align:left;color: blue !important;">
t0_B
</td>
<td style="text-align:left;color: blue !important;">
t0
</td>
<td style="text-align:left;color: blue !important;">
B
</td>
<td style="text-align:left;color: blue !important;">
A431
</td>
<td style="text-align:left;color: red !important;">
ENSG00000000419
</td>
<td style="text-align:right;color: orange !important;">
660
</td>
</tr>
<tr>
<td style="text-align:left;color: blue !important;">
Sample_2
</td>
<td style="text-align:left;color: blue !important;">
t0_B
</td>
<td style="text-align:left;color: blue !important;">
t0
</td>
<td style="text-align:left;color: blue !important;">
B
</td>
<td style="text-align:left;color: blue !important;">
A431
</td>
<td style="text-align:left;color: red !important;">
ENSG00000000457
</td>
<td style="text-align:right;color: orange !important;">
54
</td>
</tr>
</tbody>
</table>
<p>Here we will only cover the conversion from <code>wide</code> to <code>long</code>, as this is more relevant to us. For the other way around, one can look into <code>spread()</code> from the <code>tidyr</code> package.</p>
<div id="using-reshape2" class="section level2">
<h2><span class="header-section-number">4.1</span> Using reshape2</h2>
<p>By using the <code>melt()</code> function from the <strong>reshape2</strong> package we can convert the wide-formatted data into long-formatted data! Here, to combine the metadata table to the gene counts table, we will also use the <code>merge()</code> function like we did before!</p>
<pre class="r"><code>library(reshape2)
gc &lt;- read.table(&quot;data/counts_raw.txt&quot;, header = T, row.names = 1, sep = &quot;\t&quot;)
md &lt;- read.table(&quot;data/metadata.csv&quot;, header = T, sep = &quot;;&quot;)
rownames(md) &lt;- md$Sample_ID

#merging gene counts table with metadata
merged_data_wide &lt;- merge(md, t(gc), by = 0)
#removing redundant columns
merged_data_wide$Row.names &lt;- NULL

merged_data_long &lt;- melt(merged_data_wide, id.vars = c(&quot;Sample_ID&quot;,&quot;Sample_Name&quot;,&quot;Time&quot;,&quot;Replicate&quot;,&quot;Cell&quot;), variable.name = &quot;Gene&quot;, value.name = &quot;count&quot;)  
head(merged_data_long)</code></pre>
<details>
<p><summary>output</summary></p>
<pre><code>##   Sample_ID Sample_Name Time Replicate Cell            Gene count
## 1  Sample_1        t0_A   t0         A A431 ENSG00000000003   321
## 2 Sample_10       t24_A  t24         A A431 ENSG00000000003   950
## 3 Sample_11       t24_B  t24         B A431 ENSG00000000003   760
## 4 Sample_12       t24_C  t24         C A431 ENSG00000000003  1436
## 5  Sample_2        t0_B   t0         B A431 ENSG00000000003   303
## 6  Sample_3        t0_C   t0         C A431 ENSG00000000003   204</code></pre>
</details>
</div>
<div id="using-tidyr" class="section level2">
<h2><span class="header-section-number">4.2</span> Using tidyr</h2>
<p>If you are more familiar with using <code>tidyverse</code> or <code>tidyr</code> packages, you can combine tables by <code>join()</code> and then use <code>gather()</code> to make long formatted data in the same command. This is a powerful and more cleaner way of dealing with data in R.</p>
<pre class="r"><code>library(tidyverse)
gc_long &lt;- gc %&gt;%
  rownames_to_column(var = &quot;Gene&quot;) %&gt;%
  gather(Sample_ID, count, -Gene) %&gt;% 
  full_join(md, by = &quot;Sample_ID&quot;) %&gt;% 
  select(Sample_ID, everything()) %&gt;% 
  select(-c(Gene,count), c(Gene,count)) 
gc_long %&gt;%
  head(10)</code></pre>
<details>
<p><summary>output</summary></p>
<pre><code>##    Sample_ID Sample_Name Time Replicate Cell            Gene count
## 1   Sample_1        t0_A   t0         A A431 ENSG00000000003   321
## 2   Sample_1        t0_A   t0         A A431 ENSG00000000005     0
## 3   Sample_1        t0_A   t0         A A431 ENSG00000000419   696
## 4   Sample_1        t0_A   t0         A A431 ENSG00000000457    59
## 5   Sample_1        t0_A   t0         A A431 ENSG00000000460   399
## 6   Sample_1        t0_A   t0         A A431 ENSG00000000938     0
## 7   Sample_1        t0_A   t0         A A431 ENSG00000000971     0
## 8   Sample_1        t0_A   t0         A A431 ENSG00000001036   844
## 9   Sample_1        t0_A   t0         A A431 ENSG00000001084   535
## 10  Sample_1        t0_A   t0         A A431 ENSG00000001167   493</code></pre>
</details>
</div>
</div>
<div id="exercise" class="section level1">
<h1><span class="header-section-number">5</span> Exercise</h1>
<p><i class="fas fa-clipboard-list"></i> Task   Here in this exercise we have used the <code>counts_raw.txt</code> file, you can try to make similar R objects for each of the other three counts (<code>counts_filtered.txt</code>, <code>counts_vst.txt</code> and <code>counts_deseq2.txt</code>) in <code>long format</code>. So for example you would have <code>gc_filt</code>, <code>gc_vst</code> and <code>gc_deseq2</code> R objects in the end.</p>
<p><i class="fas fa-lightbulb"></i> Tip   Remember to take a look at how these files are formatted, before you import!</p>
</div>
<div id="acknowledgements" class="section level1">
<h1><span class="header-section-number">6</span> Acknowledgements</h1>
<p>Much of the data format explanations and exercises were obtained from the GitHub of <a href="https://github.com/eriqande" target="_blank">Eric C. Anderson</a></p>
</div>
<div id="getting-help" class="section level1">
<h1><span class="header-section-number">7</span> Getting help</h1>
<ul>
<li>Use <code>?function</code> to get function documentation</li>
<li>Use <code>??bla</code> to search for a function</li>
<li>Use <code>args(function)</code> to get the arguments to a function</li>
<li>Go to the package CRAN page/webpage for vignettes</li>
<li><a href="http://www.cookbook-r.com/" target="_blank">R Cookbook</a>: General purpose reference.</li>
<li><a href="https://www.statmethods.net/" target="_blank">Quick R</a>: General purpose reference.</li>
<li><a href="https://stackoverflow.com/" target="_blank">Stackoverflow</a>: Online community to find solutions to your problems.</li>
</ul>
</div>
<div id="session-info" class="section level1">
<h1><span class="header-section-number">8</span> Session info</h1>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 4.1.3 (2022-03-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.6 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
## LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so
## 
## locale:
##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   
## 
## attached base packages:
## [1] grid      stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] ggimage_0.3.1          treeio_1.18.1          ggtree_3.2.1          
##  [4] pheatmap_1.0.12        swemaps_1.0            mapdata_2.3.0         
##  [7] maps_3.4.0             gridExtra_2.3          jpeg_0.1-9            
## [10] ggpubr_0.4.0           cowplot_1.1.1          ggthemes_4.2.4        
## [13] scales_1.2.1           ggrepel_0.9.1          wesanderson_0.3.6     
## [16] forcats_0.5.2          stringr_1.4.1          purrr_0.3.5           
## [19] readr_2.1.3            tidyr_1.2.1            tibble_3.1.8          
## [22] tidyverse_1.3.2        reshape2_1.4.4         ggplot2_3.3.6         
## [25] formattable_0.2.1      kableExtra_1.3.4       dplyr_1.0.10          
## [28] lubridate_1.8.0        leaflet_2.1.1          yaml_2.3.5            
## [31] fontawesome_0.3.0.9000 captioner_2.2.3        bookdown_0.29         
## [34] knitr_1.40            
## 
## loaded via a namespace (and not attached):
##  [1] googledrive_2.0.0       colorspace_2.0-3        ggsignif_0.6.3         
##  [4] ellipsis_0.3.2          fs_1.5.2                aplot_0.1.8            
##  [7] rstudioapi_0.14         farver_2.1.1            fansi_1.0.3            
## [10] xml2_1.3.3              splines_4.1.3           cachem_1.0.6           
## [13] jsonlite_1.8.2          broom_1.0.1             dbplyr_2.2.1           
## [16] compiler_4.1.3          httr_1.4.4              backports_1.4.1        
## [19] assertthat_0.2.1        Matrix_1.5-1            fastmap_1.1.0          
## [22] lazyeval_0.2.2          gargle_1.2.1            cli_3.4.1              
## [25] leaflet.providers_1.9.0 htmltools_0.5.3         tools_4.1.3            
## [28] gtable_0.3.1            glue_1.6.2              Rcpp_1.0.9             
## [31] carData_3.0-5           cellranger_1.1.0        jquerylib_0.1.4        
## [34] vctrs_0.4.2             ape_5.6-2               svglite_2.1.0          
## [37] nlme_3.1-155            crosstalk_1.2.0         xfun_0.33              
## [40] rvest_1.0.3             lifecycle_1.0.3         rstatix_0.7.0          
## [43] googlesheets4_1.0.1     hms_1.1.2               parallel_4.1.3         
## [46] RColorBrewer_1.1-3      curl_4.3.3              ggfun_0.0.7            
## [49] yulab.utils_0.0.5       sass_0.4.2              stringi_1.7.8          
## [52] highr_0.9               tidytree_0.4.1          rlang_1.0.6            
## [55] pkgconfig_2.0.3         systemfonts_1.0.4       evaluate_0.17          
## [58] lattice_0.20-45         patchwork_1.1.2         htmlwidgets_1.5.4      
## [61] labeling_0.4.2          tidyselect_1.2.0        plyr_1.8.7             
## [64] magrittr_2.0.3          R6_2.5.1                magick_2.7.3           
## [67] generics_0.1.3          DBI_1.1.3               pillar_1.8.1           
## [70] haven_2.5.1             withr_2.5.0             mgcv_1.8-39            
## [73] abind_1.4-5             modelr_0.1.9            crayon_1.5.2           
## [76] car_3.1-0               utf8_1.2.2              tzdb_0.3.0             
## [79] rmarkdown_2.17          readxl_1.4.1            reprex_2.0.2           
## [82] digest_0.6.29           webshot_0.5.4           gridGraphics_0.5-1     
## [85] munsell_0.5.0           viridisLite_0.4.1       ggplotify_0.1.0        
## [88] bslib_0.4.0</code></pre>
<hr />
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
